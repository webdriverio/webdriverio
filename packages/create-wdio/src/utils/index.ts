
export { SUPPORTED_PACKAGE_MANAGERS } from '../constants.js'
