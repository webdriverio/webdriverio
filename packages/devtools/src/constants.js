export const DEFAULT_WIDTH = 1200
export const DEFAULT_HEIGHT = 900
export const DEFAULT_X_POSITION = 0
export const DEFAULT_Y_POSITION = 0

export const ELEMENT_KEY = 'element-6066-11e4-a52e-4f735466cecf'

export const DEFAULT_FLAGS = [
    // Disable built-in Google Translate service
    '--disable-translate',
    // Disable all chrome extensions entirely
    '--disable-extensions',
    // Disable various background network services, including extension updating,
    //   safe browsing service, upgrade detector, translate, UMA
    '--disable-background-networking',
    // Disable syncing to a Google account
    '--disable-sync',
    // Disable reporting to UMA, but allows for collection
    '--metrics-recording-only',
    // Disable installation of default apps on first run
    '--disable-default-apps',
    // Mute any audio
    '--mute-audio',
    // Skip first run wizards
    '--no-first-run',
    // chromedriver flags
    '--disable-hang-monitor',
    '--disable-prompt-on-repost',
    '--disable-web-resources',
    '--disable-client-side-phishing-detection',
    '--enable-logging',
    '--log-level=0',
    '--password-store=basic',
    '--use-mock-keychain',
    '--test-type=webdriver',
    '--force-fieldtrials=SiteIsolationExtensions/Control'
]

export const CHROME_NAMES = ['chrome', 'googlechrome', 'headlesschrome', 'google-chrome', 'chromium']
export const FIREFOX_NAMES = ['firefox', 'ff', 'mozilla', 'mozillafirefox', 'headless firefox', 'headlessfirefox']
export const EDGE_NAMES = ['edge', 'msedge', 'microsoft-edge', 'microsoftedge']

export const DEFAULTS = {
    capabilities: {
        type: 'object',
        required: true
    },
    /**
     * Level of logging verbosity
     */
    logLevel: {
        type: 'string',
        default: 'info',
        match: /(trace|debug|info|warn|error|silent)/
    },
    /**
     * Timeout for any request to the Selenium server
     */
    connectionPollInterval: {
        type: 'number',
        default: 500
    },
    /**
     * maxConnectionRetries in chrome-launcher
     */
    connectionRetryCount: {
        type: 'number',
        default: 50
    }
}

export const DEFAULT_IMPLICIT_TIMEOUT = 0
export const DEFAULT_PAGELOAD_TIMEOUT = 5 * 60 * 1000 // 5 min
export const DEFAULT_SCRIPT_TIMEOUT = 30 * 1000 // 30s

export const SUPPORTED_SELECTOR_STRATEGIES = ['css selector', 'tag name', 'xpath', 'link text', 'partial link text']
export const SERIALIZE_PROPERTY = 'data-devtoolsdriver-fetchedElement'
export const SERIALIZE_SELECTOR = `[${SERIALIZE_PROPERTY}]`
export const SERIALIZE_FLAG = '__executeElement'

export const ERROR_MESSAGES = {
    staleElement: {
        name: 'stale element reference',
        message: 'stale element reference: The element reference is stale; either the element is no longer attached to the DOM, it is not in the current frame context, or the document has been refreshed'
    }
}
