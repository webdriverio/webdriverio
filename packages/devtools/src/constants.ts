import type { DefaultOptions } from '@wdio/config'

export const DEFAULT_WIDTH = 1200
export const DEFAULT_HEIGHT = 900
export const DEFAULT_X_POSITION = 0
export const DEFAULT_Y_POSITION = 0

export const ELEMENT_KEY = 'element-6066-11e4-a52e-4f735466cecf'

// https://github.com/puppeteer/puppeteer/blob/083ea41e943e2a20014279fcfceb23c98a1e4491/src/node/Launcher.ts#L168
export const DEFAULT_FLAGS = [
    // suppresses Save Password prompt window
    '--enable-automation',
    // do not block popups
    '--disable-popup-blocking',
    // Disable all chrome extensions entirely
    '--disable-extensions',
    // Disable various background network services, including extension updating,
    //   safe browsing service, upgrade detector, translate, UMA
    '--disable-background-networking',
    '--disable-background-timer-throttling',
    '--disable-backgrounding-occluded-windows',
    // Disable syncing to a Google account
    '--disable-sync',
    // Disable reporting to UMA, but allows for collection
    '--metrics-recording-only',
    // Disable installation of default apps on first run
    '--disable-default-apps',
    // Mute any audio
    '--mute-audio',
    // Skip first run wizards
    '--no-first-run',
    '--no-default-browser-check',
    // chromedriver flags
    '--disable-hang-monitor',
    '--disable-prompt-on-repost',
    '--disable-client-side-phishing-detection',
    '--password-store=basic',
    '--use-mock-keychain',
    '--disable-component-extensions-with-background-pages',
    '--disable-breakpad',
    '--disable-dev-shm-usage',
    '--disable-ipc-flooding-protection',
    '--disable-renderer-backgrounding',
    '--force-fieldtrials=*BackgroundTracing/default/',
    '--enable-features=NetworkService,NetworkServiceInProcess',
    /**
     * `site-per-process` affects `page.frames()`, see #4471
     * `TranslateUI` disables built-in Google Translate service
     */
    '--disable-features=site-per-process,TranslateUI,BlinkGenPropertyTrees'
]

export const CHROME_NAMES = ['chrome', 'googlechrome', 'headlesschrome', 'google-chrome', 'chromium']
export const FIREFOX_NAMES = ['firefox', 'ff', 'mozilla', 'mozillafirefox', 'headless firefox', 'headlessfirefox']
export const EDGE_NAMES = ['edge', 'msedge', 'microsoft-edge', 'microsoftedge']
export const SUPPORTED_BROWSER = [...CHROME_NAMES, ...FIREFOX_NAMES, ...EDGE_NAMES]

export const BROWSER_TYPE: {
    chrome: 'chrome',
    firefox: 'firefox',
    edge: 'edge'
} = {
    chrome: 'chrome',
    firefox: 'firefox',
    edge: 'edge'
}

export const DEFAULTS: DefaultOptions<WebDriver.Options> = {
    capabilities: {
        type: 'object',
        required: true
    },
    /**
     * Level of logging verbosity
     */
    logLevel: {
        type: 'string',
        default: 'info',
        match: /(trace|debug|info|warn|error|silent)/
    },
    /**
     * maxConnectionRetries in chrome-launcher
     */
    connectionRetryCount: {
        type: 'number',
        default: 50
    }
}

export const DEFAULT_IMPLICIT_TIMEOUT = 0
export const DEFAULT_PAGELOAD_TIMEOUT = 5 * 60 * 1000 // 5 min
export const DEFAULT_SCRIPT_TIMEOUT = 30 * 1000 // 30s

export const SUPPORTED_SELECTOR_STRATEGIES = ['css selector', 'tag name', 'xpath', 'link text', 'partial link text']
export const SERIALIZE_PROPERTY = 'data-devtoolsdriver-fetchedElement'
export const SERIALIZE_FLAG = '__executeElement'

export const PPTR_LOG_PREFIX = 'puppeteer:protocol'

export const ERROR_MESSAGES = {
    staleElement: {
        name: 'stale element reference',
        message: 'stale element reference: The element reference is stale; either the element is no longer attached to the DOM, it is not in the current frame context, or the document has been refreshed'
    }
}

export const BROWSER_ERROR_MESSAGES = {
    firefoxNightly: `Only Nightly release channel is supported in Devtools/Puppeteer for Firefox. Refer to the following issue:
        https://bugzilla.mozilla.org/show_bug.cgi?id=1606604

        You can use the following link to download Firefox Nightly edition:
        https://www.mozilla.org/en-US/firefox/channel/desktop/

        Adding the following binary capability in Firefox Options is mandatory to run with Nightly edition:

        'moz:firefoxOptions': {
            binary: '/path/to/firefox'
        }

        Note: "Nightly" as a term should be present in the "Firefox Application Name" across all OS's in binary path mentioned above for this to work.`
}

export const VENDOR_PREFIX: {
    chrome: 'goog:chromeOptions',
    firefox: 'moz:firefoxOptions',
    edge: 'ms:edgeOptions'
} = {
    chrome: 'goog:chromeOptions',
    firefox: 'moz:firefoxOptions',
    edge: 'ms:edgeOptions'
}

export const CHANNEL_FIREFOX_NIGHTLY = 'nightly'
export const CHANNEL_FIREFOX_TRUNK = 'trunk'
