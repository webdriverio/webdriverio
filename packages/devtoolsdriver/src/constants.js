export const DEFAULT_WIDTH = 1200
export const DEFAULT_HEIGHT = 900
export const ELEMENT_KEY = 'element-6066-11e4-a52e-4f735466cecf'
