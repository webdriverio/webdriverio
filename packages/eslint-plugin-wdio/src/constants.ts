export const MATCHERS = [
    'toHaveTitle',
    'toHaveTitleContaining',
    'toHaveUrl',
    'toHaveUrlContaining',
    'toBeClickable',
    'toBeDisabled',
    'toBeDisplayed',
    'toBeDisplayedInViewport',
    'toBeEnabled',
    'toBeExisting',
    'toBeFocused',
    'toBeSelected',
    'toHaveAttribute',
    'toHaveAttributeContaining',
    'toHaveChildren',
    'toHaveElementClass',
    'toHaveElementClassContaining',
    'toHaveElementProperty',
    'toHaveHref',
    'toHaveHrefContaining',
    'toHaveId',
    'toHaveStyle',
    'toHaveText',
    'toHaveTextContaining',
    'toHaveValue',
    'toHaveValueContaining',
    'toBeElementsArrayOfSize',
    'toBeRequested',
    'toBeRequestedTimes',
    'toBeRequestedWith',
    'toBeRequestedWithResponse'
]
