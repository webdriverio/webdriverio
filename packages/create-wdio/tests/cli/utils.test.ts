import os from 'node:os'
import fs from 'node:fs/promises'
import { vi, test, expect, describe, beforeEach, it } from 'vitest'
import { canAccessConfigPath, missingConfigurationPrompt, parseAnswers, runConfigCommand } from '../../src/cli/utils.js'
import { createPackageJSON, npmInstall, createWDIOConfig, createWDIOScript, runAppiumInstaller, getAnswers } from '../../src/utils.js'
import { BackendChoice } from '../../src/constants.js'

import inquirer from 'inquirer'
import path from 'node:path'
import type { Questionnair } from '../../src/types.js'

vi.mock('node:fs/promises', async (orig) => ({
    ...(await orig()) as any,
    default: {
        access: vi.fn().mockRejectedValue('Yay')
    }
}))
vi.mock('inquirer')
vi.mock('../../src/utils.js', async () => {
    const actual = await vi.importActual('../../src/utils.js') as Promise<object>
    return {
        ...actual,
        getAnswers: vi.fn(),
        getProjectProps: vi.fn().mockResolvedValue({
            path: '/foo/bar',
            esmSupported: true,
            packageJson: {
                name: 'my-module'
            }
        }),
        getProjectRoot: vi.fn().mockReturnValue('/foo/bar'),
        createPackageJSON: vi.fn(),
        npmInstall: vi.fn(),
        createWDIOConfig: vi.fn(),
        createWDIOScript: vi.fn(),
        runAppiumInstaller: vi.fn()
    }
})

vi.mock('../src/install', () => ({
    installPackages: vi.fn(),
    getInstallCommand: vi.fn().mockReturnValue('npm install foo bar --save-dev')
}))
global.console.log = vi.fn()

const isUsingWindows = os.platform() === 'win32'

test('runConfigCommand', async () => {
    await runConfigCommand({ projectRootDir: '/foo/bar' } as any, 'next')
    expect(createPackageJSON).toBeCalledTimes(1)
    expect(npmInstall).toBeCalledTimes(1)
    expect(createWDIOConfig).toBeCalledTimes(1)
    expect(createWDIOScript).toBeCalledTimes(1)
    expect(runAppiumInstaller).toBeCalledTimes(1)
    expect(vi.mocked(console.log).mock.calls).toMatchSnapshot()
})

describe('canAccessConfigPath', () => {
    beforeEach(()=>{
        vi.mocked(fs.access).mockClear()
    })
    test('no extension only', async () => {
        vi.mocked(fs.access)
            .mockRejectedValueOnce(new Error('not found'))
            .mockRejectedValueOnce(new Error('not found'))
            .mockRejectedValueOnce(new Error('not found'))
            .mockResolvedValue('Yay' as any)
        expect(await canAccessConfigPath('/foo/bar')).toBe('/foo/bar.mts')
        expect(fs.access).not.toBeCalledWith('/foo/bar')
        expect(fs.access).toBeCalledWith('/foo/bar.js')
        expect(fs.access).toBeCalledWith('/foo/bar.ts')
        expect(fs.access).toBeCalledWith('/foo/bar.mjs')
        expect(fs.access).toBeCalledWith('/foo/bar.mts')
    })

    test('with full path of the existed file', async () => {
        vi.mocked(fs.access)
            .mockResolvedValue('Yay' as any)
        expect(await canAccessConfigPath('/foo/bar', '/foo/bar.ts')).toBe('/foo/bar.ts')
        expect(fs.access).toHaveBeenCalledTimes(1)
        expect(fs.access).toBeCalledWith('/foo/bar.ts')
    })

    test('with full path of the not existed file', async () => {
        vi.mocked(fs.access)
            .mockRejectedValueOnce(new Error('not found'))
            .mockRejectedValueOnce(new Error('not found'))
            .mockRejectedValueOnce(new Error('not found'))
            .mockRejectedValueOnce(new Error('not found'))
            .mockResolvedValue('Yay' as any)
        expect(await canAccessConfigPath('/foo/bar', '/foo/bar.ts')).toBe('/foo/bar.mts')
        expect(fs.access).toBeCalledWith('/foo/bar.js')
        expect(fs.access).toBeCalledWith('/foo/bar.ts')
        expect(fs.access).toBeCalledWith('/foo/bar.mjs')
        expect(fs.access).toBeCalledWith('/foo/bar.mts')
    })
})
test('missingConfigurationPrompt does not init wizard if user does not want to', async () => {
    vi.mocked(getAnswers).mockResolvedValue({
        backend: BackendChoice.Local,
        generateTestFiles: false,
        baseUrl: 'http://localhost',
        runner: '@wdio/local-runner$--$local',
        framework: '@wdio/mocha-framework$--$mocha',
        preset: '@sveltejs/vite-plugin-svelte$--$svelte',
        reporters: [],
        plugins: [],
        services: [],
        npmInstall: true
    } as any)
    const runConfigCmd = vi.fn()
    vi.mocked(inquirer.prompt).mockResolvedValue({})
    await missingConfigurationPrompt('config', 'foobar', runConfigCmd)
    expect(runConfigCmd).toBeCalledTimes(0)
})

test.skipIf(isUsingWindows)('parseAnswers', async () => {
    vi.mocked(getAnswers).mockResolvedValue({
        backend: BackendChoice.Local,
        specs: '/tmp/foobar/specs',
        pages: '/tmp/foobar/pageobjects',
        stepDefinitions: '/tmp/foobar/steps',
        generateTestFiles: true,
        usePageObjects: true,
        baseUrl: 'http://localhost',
        runner: '@wdio/local-runner$--$local',
        framework: '@wdio/mocha-framework$--$mocha',
        preset: '@sveltejs/vite-plugin-svelte$--$svelte',
        isUsingTypeScript: false,
        reporters: [
            '@wdio/spec-reporter$--$spec'
        ],
        plugins: [
            'wdio-wait-for$--$wait-for'
        ],
        services: [
            '@wdio/sauce-service$--$sauce'
        ],
        npmInstall: true
    } as any)
    const parsedAnswers = await parseAnswers(true)
    expect(parsedAnswers).toMatchSnapshot()
})

describe('Serenity/JS project generation', () => {

    const defaultAnswers: Questionnair = {
        runner: '@wdio/local-runner$--$local$--$e2e',
        framework: undefined,   // overridden in the tests
        backend: BackendChoice.Local,
        e2eEnvironment: 'web',
        browserEnvironment: ['chrome'],
        specs: '/foo/bar/specs',
        pages: '/foo/bar/pageobjects',
        generateTestFiles: true,
        usePageObjects: true,
        baseUrl: 'http://localhost',
        reporters: [],
        plugins: [],
        services: [],
        npmInstall: true,
        isUsingTypeScript: false,
    } as any

    it('marks serenityAdapter as false and destSerenityLibRootPath as blank if not using Serenity/JS', async () => {
        vi.mocked(getAnswers).mockResolvedValue({
            ...defaultAnswers,
            framework: '@wdio/mocha-framework$--$mocha',
        })
        const parsedAnswers = await parseAnswers(true)

        expect(parsedAnswers.framework).toEqual('mocha')
        expect(parsedAnswers.serenityAdapter).toEqual(false)
        expect(parsedAnswers.destSerenityLibRootPath).toEqual('')
    })

    it('adds projectName so that it can be used in templates', async () => {
        vi.mocked(getAnswers).mockResolvedValue({
            ...defaultAnswers,
            framework: '@serenity-js/webdriverio$--$@serenity-js/webdriverio$--$mocha',
            serenityLibPath: './serenity',
        })
        const parsedAnswers = await parseAnswers(true)

        expect(parsedAnswers.projectName).toEqual('my-module')
    })

    describe('serenityLibPath', () => {

        it('supports a relative path', async () => {
            vi.mocked(getAnswers).mockResolvedValue({
                ...defaultAnswers,
                framework: '@serenity-js/webdriverio$--$@serenity-js/webdriverio$--$mocha',
                serenityLibPath: './serenity',
            })
            const parsedAnswers = await parseAnswers(true)

            expect(parsedAnswers.destSerenityLibRootPath.split(path.sep))
                .toEqual(expect.arrayContaining(['foo', 'bar', 'serenity']))
        })

        it('supports an absolute path', async () => {
            vi.mocked(getAnswers).mockResolvedValue({
                ...defaultAnswers,
                framework: '@serenity-js/webdriverio$--$@serenity-js/webdriverio$--$mocha',
                serenityLibPath: '/foo/bar/src/serenity',
            })
            const parsedAnswers = await parseAnswers(true)

            expect(parsedAnswers.destSerenityLibRootPath.split(path.sep))
                .toEqual(expect.arrayContaining(['foo', 'bar', 'src', 'serenity']))
        })
    })

    describe('with Cucumber', () => {

        it('adds necessary packages', async () => {
            vi.mocked(getAnswers).mockResolvedValue({
                ...defaultAnswers,
                framework: '@serenity-js/webdriverio$--$@serenity-js/webdriverio$--$cucumber',
                specs: 'features/**/*.feature',
                stepDefinitions: 'features/step-definitions/steps'
            })
            const parsedAnswers = await parseAnswers(true)

            expect(parsedAnswers.framework).toEqual('@serenity-js/webdriverio')
            expect(parsedAnswers.serenityAdapter).toEqual('cucumber')
            expect(parsedAnswers.destSpecRootPath.split(path.sep)).toEqual(expect.arrayContaining(['foo', 'bar', 'features']))
            expect(parsedAnswers.packagesToInstall).toEqual([
                '@wdio/local-runner',
                '@serenity-js/webdriverio',
                '@cucumber/cucumber',
                '@serenity-js/assertions',
                '@serenity-js/console-reporter',
                '@serenity-js/core',
                '@serenity-js/cucumber',
                '@serenity-js/rest',
                '@serenity-js/serenity-bdd',
                '@serenity-js/web',
                'npm-failsafe',
                'rimraf',
            ])
        })

        it('supports TypeScript projects', async () => {
            vi.mocked(getAnswers).mockResolvedValue({
                ...defaultAnswers,
                framework: '@serenity-js/webdriverio$--$@serenity-js/webdriverio$--$cucumber',
                specs: 'features/**/*.feature',
                stepDefinitions: 'features/step-definitions/steps',
                isUsingTypeScript: true,
            })
            const parsedAnswers = await parseAnswers(true)

            expect(parsedAnswers.framework).toEqual('@serenity-js/webdriverio')
            expect(parsedAnswers.serenityAdapter).toEqual('cucumber')
            expect(parsedAnswers.packagesToInstall).toEqual([
                '@wdio/local-runner',
                '@serenity-js/webdriverio',
                '@cucumber/cucumber',
                '@serenity-js/assertions',
                '@serenity-js/console-reporter',
                '@serenity-js/core',
                '@serenity-js/cucumber',
                '@serenity-js/rest',
                '@serenity-js/serenity-bdd',
                '@serenity-js/web',
                '@types/node',
                'npm-failsafe',
                'rimraf',
            ])
        })
    })

    describe('with Jasmine', () => {
        it('adds necessary packages', async () => {
            vi.mocked(getAnswers).mockResolvedValue({
                ...defaultAnswers,
                framework: '@serenity-js/webdriverio$--$@serenity-js/webdriverio$--$jasmine'
            })
            const parsedAnswers = await parseAnswers(true)

            expect(parsedAnswers.framework).toEqual('@serenity-js/webdriverio')
            expect(parsedAnswers.serenityAdapter).toEqual('jasmine')
            expect(parsedAnswers.packagesToInstall).toEqual([
                '@wdio/local-runner',
                '@serenity-js/webdriverio',
                '@serenity-js/assertions',
                '@serenity-js/console-reporter',
                '@serenity-js/core',
                '@serenity-js/jasmine',
                '@serenity-js/rest',
                '@serenity-js/serenity-bdd',
                '@serenity-js/web',
                'jasmine',
                'npm-failsafe',
                'rimraf',
            ])
        })

        it('supports TypeScript projects', async () => {
            vi.mocked(getAnswers).mockResolvedValue({
                ...defaultAnswers,
                framework: '@serenity-js/webdriverio$--$@serenity-js/webdriverio$--$jasmine',
                isUsingTypeScript: true,
            })
            const parsedAnswers = await parseAnswers(true)

            expect(parsedAnswers.framework).toEqual('@serenity-js/webdriverio')
            expect(parsedAnswers.serenityAdapter).toEqual('jasmine')
            expect(parsedAnswers.packagesToInstall).toEqual([
                '@wdio/local-runner',
                '@serenity-js/webdriverio',
                '@serenity-js/assertions',
                '@serenity-js/console-reporter',
                '@serenity-js/core',
                '@serenity-js/jasmine',
                '@serenity-js/rest',
                '@serenity-js/serenity-bdd',
                '@serenity-js/web',
                '@types/jasmine',
                '@types/node',
                'jasmine',
                'npm-failsafe',
                'rimraf',
            ])
        })
    })

    describe('with Mocha', () => {

        it('adds necessary packages', async () => {
            vi.mocked(getAnswers).mockResolvedValue({
                ...defaultAnswers,
                framework: '@serenity-js/webdriverio$--$@serenity-js/webdriverio$--$mocha',
                serenityLibPath: './serenity',
            })
            const parsedAnswers = await parseAnswers(true)

            expect(parsedAnswers.framework).toEqual('@serenity-js/webdriverio')
            expect(parsedAnswers.serenityAdapter).toEqual('mocha')
            expect(parsedAnswers.packagesToInstall).toEqual([
                '@wdio/local-runner',
                '@serenity-js/webdriverio',
                '@serenity-js/assertions',
                '@serenity-js/console-reporter',
                '@serenity-js/core',
                '@serenity-js/mocha',
                '@serenity-js/rest',
                '@serenity-js/serenity-bdd',
                '@serenity-js/web',
                'mocha',
                'npm-failsafe',
                'rimraf',
            ])
        })

        it('supports TypeScript projects', async () => {
            vi.mocked(getAnswers).mockResolvedValue({
                ...defaultAnswers,
                framework: '@serenity-js/webdriverio$--$@serenity-js/webdriverio$--$mocha',
                serenityLibPath: './serenity',
                isUsingTypeScript: true,
            })
            const parsedAnswers = await parseAnswers(true)

            expect(parsedAnswers.framework).toEqual('@serenity-js/webdriverio')
            expect(parsedAnswers.serenityAdapter).toEqual('mocha')
            expect(parsedAnswers.packagesToInstall).toEqual([
                '@wdio/local-runner',
                '@serenity-js/webdriverio',
                '@serenity-js/assertions',
                '@serenity-js/console-reporter',
                '@serenity-js/core',
                '@serenity-js/mocha',
                '@serenity-js/rest',
                '@serenity-js/serenity-bdd',
                '@serenity-js/web',
                '@types/mocha',
                '@types/node',
                'mocha',
                'npm-failsafe',
                'rimraf',
            ])
        })
    })
})

test('missingConfigurationPrompt does run config if user agrees', async () => {
    vi.mocked(getAnswers).mockResolvedValue({
        backend: BackendChoice.Local,
        generateTestFiles: false,
        baseUrl: 'http://localhost',
        runner: '@wdio/local-runner$--$local',
        framework: '@wdio/mocha-framework$--$mocha',
        preset: '@sveltejs/vite-plugin-svelte$--$svelte',
        reporters: [],
        plugins: [],
        services: [],
        npmInstall: true
    } as any)
    const runConfigCmd = vi.fn()

    vi.mocked(inquirer.prompt).mockResolvedValue({ config: true })
    await missingConfigurationPrompt('config', 'foobar', runConfigCmd)
    expect(runConfigCmd).toBeCalledTimes(1)
})
