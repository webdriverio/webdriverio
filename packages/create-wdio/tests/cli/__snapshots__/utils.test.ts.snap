// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseAnswers 1`] = `
{
  "_async": "async ",
  "_await": "await ",
  "backend": "On my local machine",
  "baseUrl": "http://localhost",
  "destPageObjectRootPath": "/tmp/foobar",
  "destSerenityLibRootPath": "",
  "destSpecRootPath": "/tmp/foobar",
  "destStepRootPath": "/tmp/foobar",
  "esmSupport": true,
  "framework": "mocha",
  "generateTestFiles": true,
  "hasRootTSConfig": true,
  "installTestingLibrary": false,
  "isSync": false,
  "isUsingTypeScript": false,
  "npmInstall": true,
  "packagesToInstall": [
    "@wdio/local-runner",
    "@wdio/mocha-framework",
    "@sveltejs/vite-plugin-svelte",
    "@wdio/spec-reporter",
    "wdio-wait-for",
    "@wdio/sauce-service",
  ],
  "pages": "/tmp/foobar/pageobjects",
  "plugins": [
    "wait-for",
  ],
  "preset": "svelte",
  "projectName": "my-module",
  "projectRootDir": "/foo/bar",
  "purpose": undefined,
  "rawAnswers": {
    "backend": "On my local machine",
    "baseUrl": "http://localhost",
    "framework": "@wdio/mocha-framework$--$mocha",
    "generateTestFiles": true,
    "isUsingTypeScript": false,
    "npmInstall": true,
    "pages": "/tmp/foobar/pageobjects",
    "plugins": [
      "wdio-wait-for$--$wait-for",
    ],
    "preset": "@sveltejs/vite-plugin-svelte$--$svelte",
    "reporters": [
      "@wdio/spec-reporter$--$spec",
    ],
    "runner": "@wdio/local-runner$--$local",
    "services": [
      "@wdio/sauce-service$--$sauce",
    ],
    "specs": "/tmp/foobar/specs",
    "stepDefinitions": "/tmp/foobar/steps",
    "usePageObjects": true,
  },
  "relativePath": "",
  "reporters": [
    "spec",
  ],
  "runner": "local",
  "serenityAdapter": false,
  "services": [
    "sauce",
  ],
  "specs": "./../../tmp/foobar/specs",
  "stepDefinitions": "./../../tmp/foobar/steps",
  "tsConfigFilePath": "/tmp/foobar/tsconfig.json",
  "tsProject": "./../../tmp/foobar/tsconfig.json",
  "usePageObjects": true,
  "useSauceConnect": undefined,
  "wdioConfigPath": "/foo/bar/wdio.conf.js",
}
`;

exports[`runConfigCommand 1`] = `
[
  [
    "
",
  ],
  [
    "
ü§ñ Successfully setup project at /foo/bar üéâ

Join our Discord Community Server and instantly find answers to your issues or queries. Or just join and say hi üëã!
  üîó https://discord.webdriver.io

Visit the project on GitHub to report bugs üêõ or raise feature requests üí°:
  üîó https://github.com/webdriverio/webdriverio

To run your tests, execute:
$ cd /foo/bar
$ npm run wdio
",
  ],
]
`;
